optuna/importance/_fanova/_evaluator.py

```diff
@@ -79,67 +94,45 @@ def evaluate(
                 "`target=lambda t: t.values[0]` for the first objective value."
             )
 
-        distributions = _get_distributions(study, params)
-        if len(distributions) == 0:
-            return OrderedDict()
+        distributions = _get_distributions(study, params=params)
+        if params is None:
+            params = list(distributions.keys())
+        assert params is not None
 
         # fANOVA does not support parameter distributions with a single value.
         # However, there is no reason to calculate parameter importance in such case anyway,
         # since it will always be 0 as the parameter is constant in the objective function.
-        zero_importances = {name: 0.0 for name, dist in distributions.items() if dist.single()}
-        distributions = {name: dist for name, dist in distributions.items() if not dist.single()}
-
-        trials = []
-        for trial in _filter_nonfinite(
-            study.get_trials(deepcopy=False, states=(TrialState.COMPLETE,)), target=target
-        ):
-            if any(name not in trial.params for name in distributions.keys()):
-                continue
-            trials.append(trial)
-
-        trans = _SearchSpaceTransform(distributions, transform_log=False, transform_step=False)
-
-        n_trials = len(trials)
-        trans_params = numpy.empty((n_trials, trans.bounds.shape[0]), dtype=numpy.float64)
-        trans_values = numpy.empty(n_trials, dtype=numpy.float64)
+        non_single_distributions, single_distributions = _split_nonsingle_and_single_distributions(
+            distributions
+        )
 
-        for trial_idx, trial in enumerate(trials):
-            trans_params[trial_idx] = trans.transform(trial.params)
-            trans_values[trial_idx] = trial.value if target is None else target(trial)
+        if len(non_single_distributions) == 0:
+            return OrderedDict()
 
-        trans_bounds = trans.bounds
-        column_to_encoded_columns = trans.column_to_encoded_columns
+        trials: List[FrozenTrial] = _get_filtered_trials(study, params=params, target=target)
 
-        if trans_params.size == 0:  # `params` were given but as an empty list.
-            return OrderedDict()
+        trans = _SearchSpaceTransform(
+            non_single_distributions, transform_log=False, transform_step=False
+        )
 
-        # Many (deep) copies of the search spaces are required during the tree traversal and using
-        # Optuna distributions will create a bottleneck.
-        # Therefore, search spaces (parameter distributions) are represented by a single
-        # `numpy.ndarray`, coupled with a list of flags that indicate whether they are categorical
-        # or not.
+        trans_params: numpy.ndarray = _get_trans_params(trials, trans)
+        values: numpy.ndarray = _get_target_values(trials, target)
```

Comment:
[nits] I think target_values is more informative.

```suggestion
        target_values: numpy.ndarray = _get_target_values(trials, target)
```

Reply:
Or we may pass it to `evaluator.fit()` directly like.

```python
        evaluator.fit(
            X=_get_trans_params(trials, trans),
            y=_get_target_values(trials, target),
            search_spaces=trans.bounds,
            column_to_encoded_columns=trans.column_to_encoded_columns,
        )
```

What do you think?

Reply:
Thanks, I renamed `values` to `target_values`. PTAL. @c-bata